lib/instrumentation/trace-context.js

```diff
@@ -0,0 +1,140 @@
+  traceId: SIZES.version,
+  id: SIZES.version + SIZES.traceId,
+  flags: SIZES.version + SIZES.ids,
+
+  // Additional parentId is stored after the header content
+  parentId: SIZES.version + SIZES.ids + SIZES.flags
+}
+
+const FLAGS = {
+  requested: 0b00000001,
+  reported: 0b00000010
+}
+
+function defineLazyProp (obj, prop, fn) {
+  Object.defineProperty(obj, prop, {
+    configurable: true,
+    enumerable: true,
+    get () {
+      const value = fn()
+      Object.defineProperty(obj, prop, {
+        enumerable: true,
+        value
+      })
+      return value
+    }
+  })
+}
+
+function hexSliceFn (buffer, offset, length) {
+  return () => buffer.slice(offset, length).toString('hex')
+}
+
+function maybeHexSliceFn (buffer, offset, length) {
+  const fn = hexSliceFn(buffer, offset, length)
+  return () => {
+    const value = fn()
+    // Check for any non-zero characters to identify a valid ID
+    if (/[1-9a-f]/.test(value)) {
+      return value
+    }
+  }
+}
+
+// Move id into parentId region before generating new id
+function makeChild (buffer) {
+  buffer.copy(buffer, OFFSETS.parentId, OFFSETS.id, OFFSETS.flags)
+  crypto.randomFillSync(buffer, OFFSETS.id, SIZES.id)
+  return new TraceContext(buffer)
+}
+
+function isValidHeader (header) {
+  return /^[\da-f]{2}-[\da-f]{32}-[\da-f]{16}-[\da-f]{2}$/.test(header)
+}
+
+// NOTE: The version byte is not fully supported yet, but is not important until
+// we use the official header name rather than Elastic-APM-traceparent.
+// https://w3c.github.io/distributed-tracing/report-trace-context.html#versioning-of-traceparent
+function headerToBuffer (header) {
+  const buffer = Buffer.alloc(SIZES.all)
+  buffer.write(header.replace(/-/g, ''), 'hex')
+  return buffer
+}
+
+function resume (header) {
+  return makeChild(headerToBuffer(header))
+}
+
+function start (sampled = false) {
+  const buffer = Buffer.alloc(SIZES.all)
+  crypto.randomFillSync(buffer, OFFSETS.traceId, SIZES.ids)
+  // Apply reported and requested flags
+  if (sampled) {
+    buffer[OFFSETS.flags] |= FLAGS.reported
+    buffer[OFFSETS.flags] |= FLAGS.requested
+  }
+  return new TraceContext(buffer)
+}
+
+const bufferSymbol = Symbol('trace-context-buffer')
+
+class TraceContext {
+  constructor (buffer) {
+    this[bufferSymbol] = buffer
+    defineLazyProp(this, 'version', hexSliceFn(buffer, OFFSETS.version, OFFSETS.traceId))
+    defineLazyProp(this, 'traceId', hexSliceFn(buffer, OFFSETS.traceId, OFFSETS.id))
+    defineLazyProp(this, 'id', hexSliceFn(buffer, OFFSETS.id, OFFSETS.flags))
+    defineLazyProp(this, 'flags', hexSliceFn(buffer, OFFSETS.flags, OFFSETS.parentId))
+    defineLazyProp(this, 'parentId', maybeHexSliceFn(buffer, OFFSETS.parentId))
+  }
+
+  get requested () {
+    return !!(this[bufferSymbol][OFFSETS.flags] & FLAGS.requested)
+  }
+  get reported () {
+    return !!(this[bufferSymbol][OFFSETS.flags] & FLAGS.reported)
+  }
+  get sampled () {
+    return this.requested && this.reported
```

Comment:
I think technically `this.reported` is all that's required for it to be considered sampled.

And we should make sure (in this naive implementation) to turn on the reported flag when we instantiate a `TraceContext` based on a parent whos `requested` flag is set.

Reply:
I added something to take requested flag from the parent to add reported flag when making a child.

As for the `sampled` getter: the `requested` flag would indicate that children of that context should also be reported, so I think using both makes sense here. What are your thoughts on that?

Reply:
So there's no scenario internally in which the context could ever have the reported flag without the requested flag. It's _technically_ possible to receive that from a header though.

I added a commit to negate the reported flag on a child of a parent with `reported` and _not_ `requested`. Do you think that's correct behaviour?

Reply:
> As for the `sampled` getter: the `requested` flag would indicate that children of that context should also be reported, so I think using both makes sense here. What are your thoughts on that?

You're correct that this is true for children. But as far as I can see, this getter is only used by `Transaction.prototype.sampled`, which to my knowledge only concerns it self with the transaction it self right?

Also, technically if this was the root transaction in the trace - and it decided it should be sampled - it would not be requested, but only reported right?

> I added a commit to negate the reported flag on a child of a parent with `reported` and _not_ `requested`. Do you think that's correct behaviour?

Yes I think that's correct. @felixbarny do you agree?

Reply:
Hmm yeah you might be right. I was thinking that `requested` meant that the parent requested it. But I guess in our case the `TraceContext` actually carry two meanings: What were we asked to do by our parent and what will we ask our children to do. And in the case of a root transaction, we were technically not asked to do anything as we're the root, but we will ask our children to do this. I guess I was just thinking of this from one side